### PR TITLE
Improve git clone behavior when working with repositories with existing files

### LINKED ISSUE #2525: Allow pulling / pushing to repositories with existing content
**Is your feature request related to a problem? Please describe.**

Unable to publish or pull from repositories with existing content on initial setup of Git Sync in Insomnia Designer. When attempting to connect to a repository that already contains content such as a readme, a user will run into the error dialog detailing that they cannot pull or push depending on the action taken.

**Describe the solution you'd like**

Configure and publish to a repository with existing content.

**Describe alternatives you've considered**

Currently, the only way to connect git sync is to use a completely clean repository, or an existing insomnia designer repository.

**Additional context**

image

### LINKED ISSUE #3131: Insomnia does not work well with an existing repo
**Describe the bug**
The insomnia tool's github sync is very brittle and not well documented. After reading the documentation for Insomnia's Github sync (https://support.insomnia.rest/article/96-git-sync) we were able to get it working with a newly created GitHub repo (although this process was less than straightforward).

Our issue is it is incredibly challenging to setup Git Sync on an existing project. The majority of the projects we will be documenting in the portal already exist. We want to sync those existing repos with Insomnia. There is no best practice or documentation on how to do this.

The workaround that has been suggested is very manual and have many potential break points and just have a hacky feel:
- Create a new branch, something like "add-insomnia"
- Make an initial commit from Insomnia
- Push it to remote
- Now, we need to merge "add-insomnia" into "main" but there is no overlapping history.
- So, you need to clone via command line, then rebase "add-insomnia" onto "main"... git rebase main add-insomnia
- Then merge into main git merge add-insomnia main
- Push main git push
- For safe measure, delete your document in designer
- Clone from the dashboard

This is a request for either a better way to accomplish using Insomnia with an existing GH repo or a feature request to improve the process on the product side.

### LINKED ISSUE #3235: Can't create Collection with Git Clone
**Describe the bug**
When creating a collection with Git Clone nothing happens in the UI - only a Console error is visible via the Chrome Developers tools:

```
Uncaught (in promise) TypeError: Cannot read property '_id' of null
    at GitRepositorySettingsModal.onSubmitEdits [as _onSubmitEdits] (bundle.js:188184)
```

**To Reproduce**
1. Go to 'Dashboard'
2. Click on 'Create'
3. Click on 'Git Clone '
4. Fill in repository URL and authentication details.
5. Click on "Done"

**Expected behavior**
I expect the repository to be created as a Collection

**Screenshots**
Screenshot 2021-03-24 at 12 22 14

**Desktop (please complete the following information):**
 - OS: macOS 11.2.3 
 - Installation Method: Download (updated via in-app updating mechanism)
 - App Version: 2021.2.0

**Additional context**
Screenshot 2021-03-24 at 12 24 23

### PR DESCRIPTION
**Pull request by develohpanda**
> It should be possible to review this PR commit by commit.

#### Note
There are two ways to connect a document to a git repository. One is to **clone** the repository from the dashboard, and the other is to create a document _first_ then open it and click `setup git sync`. This PR handles both cases.

### Clone repository
Regardless of if a repository contains existing content, when cloning a repository (from the dashboard - see gif below), the user is prompted if they want to create a new document, if:
- no `.insomnia` folder is found in the root
- no `.insomnia/Workspace` folder is found in the root
- `.insomnia/Workspace` is empty

If a new document is created, the git settings are linked to that repository, allowing the user to then **commit and push** the new workspace addition. **There is never a need to create the `.insomnia` directory manually.**

These three cases are handled by the following unit tests:

image

Demo:

### "Setup Git Sync" on an existing document with a non-empty repository
Setting up git sync on an existing document should fail for any repositories that _already_ contain a workspace. An error like the following will show:

image

> In a future iteration we can automate this step by asking "would you like to clone this workspace" instead of directing the user to the dashboard.

Assuming a repository does _not_ contain a workspace, it should clone correctly and show any history attached to the repository.

image

### Fresh repository clone
Cloning a repository should fail if:
- _multiple_ files/workspaces are found in the `.insomnia/Workspace` directory
- the workspace already exists in Insomnia (already cloned)
- a `isomorphic-git` exception is thrown (eg. a network error) during clone

These three cases are handled by the following unit tests:

image

A final unit test ensures successful import if data is found:

image

> Handling malformed content in the `.insomnia` directory is **out of scope** for this PR as the expectation is users do not alter contents of the `.insomnia` directory manually, and only Insomnia reads/writes to it. There has been conversation about handling malformed data in the future.

### Visual feedback
There is currently no visual feedback, and it was quite tricky to test this ticket without seeing any visual feedback to show "something is happening" while `isomorphic-git` clones in the background. We already have support for a global loading state, controlled by the `LOAD_START` and `LOAD_STOP` actions. 

Visual feedback was introduced to the git-clone workflow, and is covered in unit tests to ensure the `LOAD_STOP` is dispatched correctly in all cases described above, however a more strict check occurs in the `afterEach` block for the git clone logic unit tests. This ensures there is always one `LOAD_START` and `LOAD_STOP` call, and that `LOAD_STOP` is always after `LOAD_START`.

image

The spinner is shown on the top left, and can be seen in the gif above.

### Refactoring
In order to achieve all of the above, the git clone workflow required the use of the document creation workflow, however both of these existed as component class methods on `WrapperHome` and were very difficult to orchestrate **and test**.

As part of this feature build, these workflows were extracted to Redux actions (using thunks for async actions) - in `packages/insomnia-app/app/ui/redux/modules/(workspace|git).js`, consumed by `WrapperHome` and `GitSyncDropdown`.

The reason for extracting these workflows to Redux is two-fold - firstly it is easier to orchestrate how the user flows through the different workflows without the need for code-duplication, and secondly it allows us to test each and every branch in a more systematic manner. The above files now have close to 100% test coverage (in terms of lines), which doesn't mean it is perfect but does mean we have more confidence in what is happening in during each expected/edge case of the git-clone and workspace-creation workflows.

It's important to know what _isn't_ tested. These unit tests do not actually clone from a git repository. The `isomorphic-git` clone method is mocked; these unit tests assume that after cloning successfully the underlying data is in a particular state, after which our business logic kicks in. We are only testing out business logic and user flows.

### QA
- [x] Clone a repo with a `.insomnia` directory  
- [x] Clone a repo with no `.insomnia` directory (but does have other files)
- [x] Clone a truly blank repo
- [x] Clone the same repo twice (should fail the second time)
- [x] Setup git sync with a repo containing a `.insomnia` directory
- [x] Setup git sync with a repo containing no `.insomnia` directory (but does have other files)
- [x] Setup git sync with a truly blank repo
- [x] Setup git sync with a repo containing at least one `.insomnia/Workspace/wrk_*.json` file (should fail and tell you to clone, see screenshot above)

### TODO
- [x] A few more unit tests

Fixes INS-522, Closes #3131, Closes #2525, Closes #3235, Fixes INS-573

### CODE REVIEW

**Review by sonicyeti — APPROVED** [review]
Very smooth and feels peppier as well.

File: packages/insomnia-app/app/ui/redux/modules/workspace.js [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+};
+
+export type GitCloneWorkspaceCallback = ({ createFsPlugin: () => Object }) => void;
+
+export const gitCloneWorkspace: GitCloneWorkspaceCallback = ({ createFsPlugin }) => {
+  return dispatch => {
+    dispatch(loadStart());
+
+    // This is a huge flow and we don't really have anywhere to put something like this. I guess
+    // it's fine here for now (?)
```
Comment (develohpanda) [inline packages/insomnia-app/app/ui/redux/modules/workspace.js]: Remove comment. This flow can be broken down now but triggering through Redux seems to be a good resting place

File: packages/insomnia-app/app/models/helpers/__tests__/git-repository-operations.test.js [hunk 2]
```diff
@@ -0,0 +1,62 @@
+// @flow
+
+import { globalBeforeEach } from '../../../__jest__/before-each';
+import * as models from '../../index';
+import { createGitRepository, deleteGitRepository } from '../git-repository-operations';
+
+describe('gitRepositoryOperations', () => {
```
Comment (develohpanda) [inline packages/insomnia-app/app/models/helpers/__tests__/git-repository-operations.test.js:7]: image

File: packages/insomnia-app/app/ui/redux/modules/__tests__/git.test.js [hunk 3]
```diff
@@ -0,0 +1,463 @@
+// @flow
```
Comment (develohpanda) [inline packages/insomnia-app/app/ui/redux/modules/__tests__/git.test.js:1]: image

File: packages/insomnia-app/app/ui/redux/modules/__tests__/workspace.test.js [hunk 4]
```diff
@@ -0,0 +1,79 @@
+// @flow
```
Comment (develohpanda) [inline packages/insomnia-app/app/ui/redux/modules/__tests__/workspace.test.js:1]: image

**Review by develohpanda — COMMENTED** [review]
Screenshots of tests to make it easier to read

**Review by sonicyeti — APPROVED** [review]
Awesome!

**Review by dimitropoulos — APPROVED** [review]
just functionally tested this with an existing repo with no files, as well as an existing repo with a .insomnia directory with only a `/.insomnia/test.txt` file, as well as an existing repo with a full-blown .insomnia directory.

**Comment by RimpiSarkar** [comment]
Is there any way to push data into local git branches, not directly to the repo

**Comment by develohpanda** [comment]
> Is there any way to push data into local git branches, not directly to the repo

Hi @RimpiSarkar, unfortunately this is not possible, it needs to be linked up with a remote repository right now
